File: Aiyagari1994/Python/Functions/fnOptaprime.py
from typing import Tuple
import numpy as np
from scipy.optimize import minimize_scalar # type: ignore
from .fnHHobj import fnHHobj

def fnOptaprime(
        objargs: Tuple,
        amin: float,
        amax: float
        ) -> float:
    """
    Minimize Scalar Function With Constraints

    Args: 
        objargs: Tuple of arguments for fnHHobj (beta, budget, grida, Na, expvf)
        amin: Lower bound for the minimization
        amax: Upper bound for the minimization

    Returns:
        float: Optimal value that minimizes objective function     
    """    
    result = minimize_scalar(
        lambda x: fnHHobj(x, *objargs), 
        bounds=[amin, amax], 
        method='bounded', 
        options={'xatol': 1e-12, 'maxiter': 2000}
    )
    return result.x

File: Aiyagari1994/Matlab/Functions/fnOptaprime.m
%% Optimization (Minimizaiton) Solver %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%   Find the Minimizer of a Function    
%
%   Args:
%       funopts: cellarray for objective function options {beta, budget, grid, Ngrid, expvf}
%       lb: lower bound of minimizer
%       ub: upper bound of minimizer
%
%   Returns:
%       aprime: (scalar) minimizer of objective function
%
function [aprime] = fnOptaprime(funopts, lb, ub)
    [beta, budget, grid, Ngrid, expvf] = deal(funopts{:});
    options = optimset('Display', 'off', 'MaxIter', 2000, 'tolX', 1e-10);
    [aprime, ~] = fminbnd((@(x) fnHHobj(x, beta, budget, grid, Ngrid, expvf)), lb, ub, options);
end